### PR TITLE
Changed window position generation to be safer for height

### PR DESCRIPTION
**Pull request by Honoo**
Creates separate safety bounds for width and height, the height bound being far more conservative to ensure most of the window is within the display. The width bound tries to ensure the windows are scattered across the page so they aren't all bunched together.

Reason for change: While using Autocompaste, a window would sometimes be generated quite close to the bottom of the display. If I used Ctrl+F to search for text that was hidden below the display, it would jump to that section of the text, as shown in this screenshot:

This led to multiple problems. The first was not being able to go back to the text editor window (the scrollbar doesn't scroll back up there), and the second was a problem with dragging windows:

I would try to drag a window upwards, and what would happen is that the header would disappear beyond the upper limit of the display, as shown in the screenshot. Because I was dragging that window, it would be permanently selected since my cursor wouldn't be able to reach the header anymore to deselect it.

tl;dr doesn't strictly fix the problems, but it makes it far less likely that they would be encountered at all.

### CODE REVIEW

**Comment by yangshun** [comment]
Thanks Su Gi! Sorry I merged this so late. Didn't want to confuse the students with a new commit in master. Hope the assignment turned out well for you :smile:
